nixos/modules/services/networking/sozu.nix

```diff
@@ -0,0 +1,303 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.sozu;
+  format = pkgs.formats.toml {};
+
+  settings = format.generate "settings.toml" cfg.settings;
+
+  configFile = pkgs.writeText "config.toml" ''
+    ${fileContents settings}
+
+    ${optionalString (cfg.extraConfig != null) cfg.extraConfig}
+  '';
```

Comment:
Hi! Thanks for doing this, but community consensus is that, in most cases, we shouldn't attempt to describe config options unless doing so is required to make other parts of the module work.

https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md

Reply:
I tried my best to avoid doing what I did in the end, but couldn't manage to completely stick with the structural `settings` pattern. As this took me much longer than expected, I decided to — before wasting even more time — just create the PR and see what people say.

The main issue I had was making options _completely_ optional, which I was told should be done with `types.nullOr types.<expected type>`. This was impossible due to the apparent nature of the TOML formatter which doesn't accept null-values, but also seems to need a declared value for every defined option (either a default value or one set by the user, but it _has_ to be there).
I think there were some more issues that I can't accurately remember at the moment; they'll probably come up eventually.

There most likely is a very simple way to do this that I just couldn't figure out.
I did try asking around on Discord, but didn't really manage to get an answer.

Reply:
Hmm. In general, most of Sozu's options shouldn't be declared in the NixOS config, so this shouldn't be a problem? I might be misunderstanding but the entire settings option should be typed as a `types.submodule { freeformType = format.type; };` and left at that.

Reply:
Oh, wow... I've tried this before, but completely misunderstood the purpose of `freeformType`; this makes so much more sense.
I'll get right back to making things better in a bit. Big thanks for the clarification!

Reply:
Right, so, this works wonderfully - but I'm not sure how to make the documentation for what's not declared in `sozu.nix` visible & clear to the user.
How do you think should I approach this?

For example:
```nix
{
  services.sozu = {
    # Has description
    enable = true;

    # Has description
    settings = {

      # Does not have description, since it's not a declared/strictly required option
      metrics = {
        address = "<whatever>";
      };

      # Does not have description, since it's not a declared/strictly required option
      listeners = [
        { protocol = "http"; } {protocol = "https"; } 
      ];
    };
  };
}
```
This works, and is needed to actually set listeners and yada yada up, but that's not at all clear to the user of this service.
Where should I describe this process, if it's freeform?

Reply:
Yes, but the toml documentation does in no way suffice for a new nix user. I can say this from experience; I just know how to do it because I wrote this module and I had to look into it - which is why I feel obligated to make the process easier for those using said module. Isn't that what this is ultimately supposed to do?

Reply:
You should write an abstraction in the form of an `option` instead of having an `extraConfig` option. This situation is (yet another way) where NixOS can shine and add so much value to the sysadmin.